library/src/commonMain/kotlin/com/badoo/reaktive/observable/DoOnBefore.kt

```diff
@@ -0,0 +1,121 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.disposable.wrap
+import com.badoo.reaktive.utils.lock.newLock
+import com.badoo.reaktive.utils.lock.synchronized
+
+fun <T> Observable<T>.doOnBeforeSubscribe(action: (Disposable) -> Unit): Observable<T> =
+    observable { observer ->
+        subscribeSafe(
+            object : ObservableObserver<T> by observer {
+                override fun onSubscribe(disposable: Disposable) {
+                    action(disposable)
+                    observer.onSubscribe(disposable)
+                }
+            }
+        )
+    }
+
+fun <T> Observable<T>.doOnBeforeNext(consumer: (T) -> Unit): Observable<T> =
```

Comment:
Couldn't this just be doOnNext? Would the user ever care about the "before" part on most of these methods.

Reply:
I wanted to stress that it's exactly "before". Later may add events like onAfterTerminate (like in RxJava).